packages/office-ui-fabric-react/src/components/DocumentCard/DocumentCardTitle.base.tsx

```diff
@@ -95,50 +100,46 @@ export class DocumentCardTitleBase extends BaseComponent<IDocumentCardTitleProps
     return documentCardTitle;
   }
 
-  private _startTruncation = (props: IDocumentCardTitleProps): void => {
+  // Truncate logic here way can't handle the case that chars with different widths are mixed very well, let _shrinkTitle take care of that.
+  private _truncateTitle = (props: IDocumentCardTitleProps): void => {
+    if (!this.state.needMeasurement) {
+      return;
+    }
+
     const originalTitle = props.title;
-    this._isTruncated = false;
-    const miniLength = props.showAsSecondaryTitle ? TRUNCATION_MINI_LENGTH_SECONDARY : TRUNCATION_MINIMUM_LENGTH;
-    const maxLength = props.showAsSecondaryTitle ? TRUNCATION_MAX_LENGTH_SECONDARY : TRUNCATION_MAXIMUM_LENGTH;
-
-    // If the title is really short, there's no need to truncate it
-    if (originalTitle && originalTitle.length >= miniLength) {
-      // Break the text into two pieces for assembly later
-      if (originalTitle.length > maxLength) {
-        // The text is really long, so we can take a chunk out of the middle so the two pieces combine for the maximum length
-        this._isTruncated = true;
-        this.setState({
-          truncatedTitleFirstPiece: originalTitle.slice(0, maxLength / 2 + TRUNCATION_FIRST_PIECE_LONGER_BY),
-          truncatedTitleSecondPiece: originalTitle.slice(originalTitle.length - (maxLength / 2 - TRUNCATION_FIRST_PIECE_LONGER_BY))
-        });
-      } else {
-        // The text is not so long, so we'll just break it into two pieces
-        this.setState({
-          truncatedTitleFirstPiece: originalTitle.slice(0, Math.ceil(originalTitle.length / 2) + TRUNCATION_FIRST_PIECE_LONGER_BY),
-          truncatedTitleSecondPiece: originalTitle.slice(
-            originalTitle.length - Math.floor(originalTitle.length / 2) + TRUNCATION_FIRST_PIECE_LONGER_BY
-          )
-        });
+    const element: HTMLDivElement | null = this._measureTitleElement.current;
+    if (element) {
+      const style: CSSStyleDeclaration = getComputedStyle(element);
```

Comment:
Perf problem here.

Reply:
There has to be a better approach; I was wondering if there could be something like:

Make a dedicated TruncatedText component.
As a prop, take in the `lineHeight` and `maxRows`.
`max-height` is set to `lineHeight * maxRows` with `overflow:hidden`.
On mount, `requestAnimationFrame`.
In the RAF callback, get the clientHeight of the container and the inner text.
If inner text bigger, render using an alternative "truncated style", which renders wrapped text, with a `max-height` of `maxRows - 1`, where the last row is renders the full text, aligned right, and a `...` glyph is overlay on top of the text on the left side.

Reply:
@dzearing it's hard to get `lineHeight` and `maxRows`

Reply:
@dzearing what if the maxRows is one row...

Reply:
requestAnimationFrame> addressed, thanks